fill in some key concepts (#237)
GlareDB/glaredb.github.io: introduction/key-concepts-in-glaredb.md

GlareDB shifts the dynamic by recognizing data is accessible where it is.
However, it is often the case that the same sources are accessed frequently and
benefit from being recognized as known sources.

This is where external data sources shine: tables and entire databases can be
added to GlareDB. By cataloging your data with GlareDB, it is accessible
wherever you use GlareDB, including applications, notebooks, data dashboards
and more.

The following adds an entire Postgres database to GlareDB's catalog, so that the
data can be accessed using `my_pg.schema.table`. Note that the data is not
imported/stored in GlareDB: it still resides in the external system and can
be accessed on-demand.

```sql
CREATE EXTERNAL DATABASE my_pg FROM postgres OPTIONS (
    host = 'my_host',
    port = '5432',
    user = 'postgres',
    password = 'postgres'
)
```

A heuristic is to use table functions for:

- Querying local, scratch or exploratory data
- One-offs or infrequent queries

and add external data sources for the remaining cases.

## Native Data Storage

In addition to reading files and external data sources, GlareDB has native

> storage. Tables can be created and data can be inserted, updated and delete as

Author: ```suggestion
storage. Tables can be created and data can be inserted, updated, and deleted as
```

and using cloud compute. Data can be accessed and stored in cloud and shared
with your team. When using [GlareDB Cloud], your can access your data in all of
your applications.

## Remote Execution

GlareDB is compatible with the Postgres protocol, meaning that tools used to
connect to Postgres can be used with GlareDB. For example, `lib/pq` in a Go
application.

When connecting via the Postgres protocol, all execution occurs in GlareDB. For
example if using [GlareDB Cloud], queries will not be hybrid: all execution
occurs in the cloud.

## Serverless Native

GlareDB has always been designed to be serverless. Storage and compute are
independently scalable and compute nodes are multi-tenant. Any GlareDB process
in the cloud can be allocated for units of work, meaning that queries can be
partitioned and distributed in parallel.

By designing the system this way, our billing system is consumption-based and
transparent. We bill for bytes processed and bytes stored, rather than resource
or time-based metrics. Billing on these metrics means that queries can get
faster, spanning multiple worker nodes and various compute resources without
the end-user incurring any change in cost!

> For more information, read our [announcement post on GlareDB Pro]

Author: ```suggestion
For more information, read our [announcement post on GlareDB Pro].
```

---
title: Key Concepts
layout: default
nav_order: 1
parent: Introduction / Getting Started
---

# Key Concepts in GlareDB

## Table Functions

> A core tenant of GlareDB is that [All Data are SQL Addressable], even data not

Reviewer: ```suggestion
A core tenet of GlareDB is that [All Data are SQL Addressable], even data not
```

## External Data Sources

In a typical "modern data stack", data is extracted and transformed from various
sources and loaded into an analytics warehouse. An underlying assumption is
analytics should own storage and infrastructure, often at great complexity.

> GlareDB shifts the dynamic by recognizing data is accessible where it is.
> However, it is often the case that the same sources are accessed frequently and
> benefit from being recognized as known sources.

Reviewer 2: ```suggestion
GlareDB provides access to data where it is, without requiring a move.
Even though data is more dynamic, having data sources configured in GlareDB 
makes it easier to access common data sources and share configuration between
users.
```

## Native Data Storage

In addition to reading files and external data sources, GlareDB has native

> storage. Tables can be created and data can be inserted, updated and delete as

Reviewer 2: ```suggestion
In addition to reading files and external data sources, GlareDB its own (native)
storage. You can create tables, and then insert, update, and delete data as needed, as in:
```

needed.

```sql
CREATE TABLE my_table (
    a_column TEXT
);

INSERT INTO my_table VALUES ('hello world');

DELETE FROM my_table;
```

## Hybrid Execution

Often, workloads with GlareDB start locally in the [CLI] or with language
bindings such as [Python] and [Node.js]. In these workloads, GlareDB runs on
a local machine either entirely in-memory with no persistence after the process
exits, or with local persistence.

[GlareDB Cloud] extends and scales these workflows. By connecting local clients
to [GlareDB Cloud], queries are partitioned and optimized to run both in-process
and using cloud compute. Data can be accessed and stored in cloud and shared
with your team. When using [GlareDB Cloud], your can access your data in all of
your applications.

## Remote Execution

GlareDB is compatible with the Postgres protocol, meaning that tools used to
connect to Postgres can be used with GlareDB. For example, `lib/pq` in a Go
application.

When connecting via the Postgres protocol, all execution occurs in GlareDB. For
example if using [GlareDB Cloud], queries will not be hybrid: all execution
occurs in the cloud.

## Serverless Native

> GlareDB has always been designed to be serverless. Storage and compute are
> independently scalable and compute nodes are multi-tenant. Any GlareDB process
> in the cloud can be allocated for units of work, meaning that queries can be
> partitioned and distributed in parallel.

Reviewer 2: ```suggestion
GlareDB is built from the bottom up with serverless operation and delivery in mind.
Storage and compute are separated and can be scaled independently, and all billing 
is based on data processed (not time!) so users only pay for the resources they use.
```
